torch/csrc/jit/codegen/cuda/kernel_cache.cpp

```diff
@@ -496,14 +498,25 @@ std::vector<at::Tensor> FusionKernelRuntime::runWithInput(
       if (iter != runtime_workspace_.tensor_map.end()) {
         fusion_outputs.push_back(iter->second);
       } else {
+        bool empty_type_check = output->getDataType().has_value() &&
+                output->getDataType().value() == DataType::Float;
+
+        // Only support two cases of empty tensor here, since 
+        //   this is hot path.
+        auto out_tv = output->as<TensorView>();
+        bool empty_tensor_check = out_tv->nDims() == 0;
+        if(!empty_tensor_check && out_tv->nDims() == 1){
+          empty_tensor_check = out_tv->axis(0)->extent()->isZeroInt();
+        }
+
         // This is the check for an empty tensor;
         TORCH_INTERNAL_ASSERT(
-            output->as<TensorView>()->nDims() == 0 &&
-                output->getDataType().has_value() &&
-                output->getDataType().value() == DataType::Float,
+            empty_tensor_check && empty_type_check,
             "Non empty tensor cannot be found at tensor_map in ",
             __FUNCTION__);
-        fusion_outputs.emplace_back(at::Tensor());
+        const auto tensor_options =
+          at::TensorOptions().dtype(at::kFloat).device(device);
+        fusion_outputs.emplace_back(at::empty({0},tensor_options));
```

Comment:
@jjsjann123 @rdspring1 some suggestions needed here on `test_batch_norm_half`

2 minor details both related to the empty tensors in inference mode:

1.  [https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/codegen/cuda/ops/normalization.cpp#L343-L344](https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/codegen/cuda/ops/normalization.cpp#L343-L344)
The op seems to pass a dim-1 size-0 tensor to represent empty while the check here expects dim-0. I just changed here to support both cases but maybe we could just consistently use 1. 

2. It looks like ShapeAnalysis in torchscript does not like passing to batchnorm backward just an at::Tensor(), though I wasn't sure if we need to test backward when not tracking running stats. A dim-1 size-0 tensor works, but we'd need to actually allocate a tensor in here.

Reply:
Hmmm, wondering how this slips through our python BN tests. :confused: 

1. good catch with the inconsistencies! Christian wanted a refactor on us generating size-0 tensors. I'll follow up with that and clean things up afterwards. I don't know if there's a canonical way to init size-0 tensors, dim-0 size-0 sounds like a reasonable representation to me, but I think in aten, I've seen all crazy representations (for scalar tensors), so maybe we can put a real check that iterates through all dimensions?
i.e. `TensorView::is_zero_tensor()` and `TensorView::is_scalar_tensor()` Any thoughts? (we don't have to do it in this PR though :heart: )

2. Is it just complaining about missing other tensor options?
> I wasn't sure if we need to test backward when not tracking running stats
There's some CI running that, IIRC, so yes we still do. (And I thought it worked at some point).
Same as above, dim-1 size-0 maybe fine here. Would you mind putting a TODO, so we'll clean it up later as well.

Reply:
Sure I can add a TODO here. 

Regarding 2:
This is the error message if we use a `at::Tensor` as the output. Looks like an arg matching problem at a glance. 

```
torch/csrc/jit/ir/alias_analysis.cpp":611, please report a bug to PyTorch. We don't have an op for aten::_batch_norm_impl_index_backward but it isn't a special case.  Argument types: int, Tensor, Tensor, Tensor, Tensor, Tensor, Tensor, Tensor, bool, float, bool[], Tensor, 

Candidates:
        aten::_batch_norm_impl_index_backward(int impl_index, Tensor input, Tensor grad_output, Tensor? weight, Tensor? running_mean, Tensor? running_var, Tensor? save_mean, Tensor? save_var_transform, bool train, float eps, bool[3] output_mask, Tensor reservedSpace) -> (Tensor, Tensor, Tensor)

```

Reply:
It's the aliasDB entries.
hmmm, that looks strange... as the signature which it complaints seem to match the signature from the candidates, except it's expecting an optional tensor but we are giving a size-0 tensor. No idea why that would trigger an error. :shrug:

Reply:
I think it's about not having a dtype for the tensor, and Tensor[Undefined] != Tensor in here:
[https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/ir/ir.cpp#L1032-L1034](https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/ir/ir.cpp#L1032-L1034)